diff --git a/llama.cpp b/llama.cpp
index 414d390e8..1250dd8ec 100644
--- a/llama.cpp
+++ b/llama.cpp
@@ -1612,15 +1612,6 @@ struct llama_mlock {
         char* errmsg = std::strerror(errno);
         bool suggest = (errno == ENOMEM);
 
-        // Check if the resource limit is fine after all
-        struct rlimit lock_limit;
-        if (suggest && getrlimit(RLIMIT_MEMLOCK, &lock_limit)) {
-            suggest = false;
-        }
-        if (suggest && (lock_limit.rlim_max > lock_limit.rlim_cur + size)) {
-            suggest = false;
-        }
-
         LLAMA_LOG_WARN("warning: failed to mlock %zu-byte buffer (after previously locking %zu bytes): %s\n%s",
                 size, this->size, errmsg, suggest ? MLOCK_SUGGESTION : "");
         return false;
