diff --git a/CMakeLists.txt b/CMakeLists.txt
index 8fdef01..9ba607b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -9,36 +9,18 @@ set(LLAMA_STATIC ON)
 set(LLAMA_OPENMP OFF)
 
 set(LLAMA_CPP_DIR "${CMAKE_CURRENT_SOURCE_DIR}/vendor/llama.cpp")
-set(SQLITE_AMALGAMATION_DIR "${CMAKE_CURRENT_SOURCE_DIR}/vendor/sqlite")
-
-include(ExternalProject)
-
-set(SQLITE_VERSION 3450300)
-set(SQLITE_YEAR 2024)
-set(SQLITE_URL https://www.sqlite.org/${SQLITE_YEAR}/sqlite-amalgamation-${SQLITE_VERSION}.zip)
-ExternalProject_Add(sqlite_amalgamation
-    URL             ${SQLITE_URL}
-    DOWNLOAD_DIR    ${CMAKE_BINARY_DIR}/downloads
-    SOURCE_DIR      ${SQLITE_AMALGAMATION_DIR}
-    CONFIGURE_COMMAND   ""
-    BUILD_COMMAND       ""
-    INSTALL_COMMAND     ""
-)
-
 
 add_subdirectory(${LLAMA_CPP_DIR} ${CMAKE_BINARY_DIR}/llama.cpp)
 
-include_directories(${SQLITE_AMALGAMATION_DIR})
+include_directories("@DEP_INSTALL_PREFIX@/sqlite3/include")
 
 add_library(sqlite_lembed SHARED sqlite-lembed.c)
-add_dependencies(sqlite_lembed sqlite_amalgamation)
 target_link_libraries(sqlite_lembed ggml_static llama)
 target_include_directories(sqlite_lembed PRIVATE ${LLAMA_CPP_DIR})
 set_target_properties(sqlite_lembed PROPERTIES PREFIX "")
 set_target_properties(sqlite_lembed PROPERTIES OUTPUT_NAME "lembed0")
 
 add_library(sqlite_lembed_static STATIC sqlite-lembed.c)
-add_dependencies(sqlite_lembed_static sqlite_amalgamation)
 target_link_libraries(sqlite_lembed_static ggml_static llama)
 target_include_directories(sqlite_lembed_static PRIVATE ${LLAMA_CPP_DIR})
 target_compile_definitions(sqlite_lembed_static PRIVATE SQLITE_CORE)
diff --git a/Makefile b/Makefile
index b11db35..247970f 100644
--- a/Makefile
+++ b/Makefile
@@ -69,28 +69,21 @@ $(BUILD_DIR):
 	cmake -B $@ $(LLAMA_CMAKE_FLAGS)
 
 
-ifdef CONFIG_WINDOWS
-ifdef release
-BUILT_LOADABLE_PATH=$(BUILD_DIR)/Release/lembed0.$(LOADABLE_EXTENSION)
-EXTRA_CMAKE_BUILD=--config Release
-else
-BUILT_LOADABLE_PATH=$(BUILD_DIR)/Debug/lembed0.$(LOADABLE_EXTENSION)
-endif
-else
 BUILT_LOADABLE_PATH=$(BUILD_DIR)/lembed0.$(LOADABLE_EXTENSION)
-endif
+BUILT_STATIC_PATH=$(BUILD_DIR)/libsqlite_lembed0.a
 
 $(TARGET_LOADABLE): sqlite-lembed.c sqlite-lembed.h $(BUILD_DIR) $(prefix)
 	cmake --build $(BUILD_DIR) -t sqlite_lembed $(EXTRA_CMAKE_BUILD)
 	ls $(BUILD_DIR)
 	cp $(BUILT_LOADABLE_PATH) $@
 
+$(TARGET_STATIC): sqlite-lembed.c sqlite-lembed.h $(BUILD_DIR) $(prefix)
+	cmake --build $(BUILD_DIR) -t sqlite_lembed_static $(EXTRA_CMAKE_BUILD)
+	ls $(BUILD_DIR)
+	cp $(BUILT_STATIC_PATH) $@
 
-sqlite-lembed.h: sqlite-lembed.h.tmpl VERSION
-	VERSION=$(shell cat VERSION) \
-	DATE=$(shell date -r VERSION +'%FT%TZ%z') \
-	SOURCE=$(shell git log -n 1 --pretty=format:%H -- VERSION) \
-	envsubst < $< > $@
+sqlite-lembed.h: sqlite-lembed.h.tmpl
+	cp $< $@
 
 MODELS_DIR=$(prefix)/.models
 
