diff --git a/CMakeLists.txt b/CMakeLists.txt
index da8a010..fae63f8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,38 +8,17 @@ set(LLAMA_METAL OFF)
 set(LLAMA_STATIC ON)
 set(LLAMA_OPENMP OFF)
 
-set(LLAMA_CPP_DIR "${CMAKE_CURRENT_SOURCE_DIR}/vendor/llama.cpp")
-set(SQLITE_AMALGAMATION_DIR "${CMAKE_CURRENT_SOURCE_DIR}/vendor/sqlite")
-
-include(ExternalProject)
-
-set(SQLITE_VERSION 3450300)
-set(SQLITE_YEAR 2024)
-set(SQLITE_URL https://www.sqlite.org/${SQLITE_YEAR}/sqlite-amalgamation-${SQLITE_VERSION}.zip)
-ExternalProject_Add(sqlite_amalgamation
-    URL             ${SQLITE_URL}
-    DOWNLOAD_DIR    ${CMAKE_BINARY_DIR}/downloads
-    SOURCE_DIR      ${SQLITE_AMALGAMATION_DIR}
-    CONFIGURE_COMMAND   ""
-    BUILD_COMMAND       ""
-    INSTALL_COMMAND     ""
-)
-
-
-add_subdirectory(${LLAMA_CPP_DIR} ${CMAKE_BINARY_DIR}/llama.cpp)
-
-include_directories(${SQLITE_AMALGAMATION_DIR})
+set(DEP_INSTALL_PREFIX "" CACHE PATH "Where other dependencies are installed")
+include_directories("${DEP_INSTALL_PREFIX}/sqlite3/include")
+include_directories("${DEP_INSTALL_PREFIX}/llama.cpp/include")
+link_directories("${DEP_INSTALL_PREFIX}/llama.cpp/lib")
 
 add_library(sqlite_lembed SHARED sqlite-lembed.c)
-add_dependencies(sqlite_lembed sqlite_amalgamation)
-target_link_libraries(sqlite_lembed ggml llama)
-target_include_directories(sqlite_lembed PRIVATE ${LLAMA_CPP_DIR})
+target_link_libraries(sqlite_lembed ggml-cpu ggml-base ggml llama)
 set_target_properties(sqlite_lembed PROPERTIES PREFIX "")
 set_target_properties(sqlite_lembed PROPERTIES OUTPUT_NAME "lembed0")
 
 add_library(sqlite_lembed_static STATIC sqlite-lembed.c)
-add_dependencies(sqlite_lembed_static sqlite_amalgamation)
 target_link_libraries(sqlite_lembed_static ggml_static llama)
-target_include_directories(sqlite_lembed_static PRIVATE ${LLAMA_CPP_DIR})
 target_compile_definitions(sqlite_lembed_static PRIVATE SQLITE_CORE)
 set_target_properties(sqlite_lembed_static PROPERTIES OUTPUT_NAME "sqlite_lembed0")
diff --git a/Makefile b/Makefile
index 5a2008e..8c34c7b 100644
--- a/Makefile
+++ b/Makefile
@@ -73,31 +73,24 @@ BUILD_DIR=$(prefix)/.build
 
 $(BUILD_DIR):
 	rm -rf @ || true
-	cmake -B $@ $(LLAMA_CMAKE_FLAGS)
+	cmake -B $@ $(LLAMA_CMAKE_FLAGS) -DDEP_INSTALL_PREFIX="@DEP_INSTALL_PREFIX@"
 
 
-ifdef CONFIG_WINDOWS
-ifdef release
-BUILT_LOADABLE_PATH=$(BUILD_DIR)/Release/lembed0.$(LOADABLE_EXTENSION)
-EXTRA_CMAKE_BUILD=--config Release
-else
-BUILT_LOADABLE_PATH=$(BUILD_DIR)/Debug/lembed0.$(LOADABLE_EXTENSION)
-endif
-else
 BUILT_LOADABLE_PATH=$(BUILD_DIR)/lembed0.$(LOADABLE_EXTENSION)
-endif
+BUILT_STATIC_PATH=$(BUILD_DIR)/libsqlite_lembed0.a
 
 $(TARGET_LOADABLE): sqlite-lembed.c sqlite-lembed.h $(BUILD_DIR) $(prefix)
 	cmake --build $(BUILD_DIR) -t sqlite_lembed $(EXTRA_CMAKE_BUILD)
 	ls $(BUILD_DIR)
 	cp $(BUILT_LOADABLE_PATH) $@
 
+$(TARGET_STATIC): sqlite-lembed.c sqlite-lembed.h $(BUILD_DIR) $(prefix)
+	cmake --build $(BUILD_DIR) -t sqlite_lembed_static $(EXTRA_CMAKE_BUILD)
+	ls $(BUILD_DIR)
+	cp $(BUILT_STATIC_PATH) $@
 
-sqlite-lembed.h: sqlite-lembed.h.tmpl VERSION
-	VERSION=$(shell cat VERSION) \
-	DATE=$(shell date -r VERSION +'%FT%TZ%z') \
-	SOURCE=$(shell git log -n 1 --pretty=format:%H -- VERSION) \
-	envsubst < $< > $@
+sqlite-lembed.h: sqlite-lembed.h.tmpl
+	cp $< $@
 
 MODELS_DIR=$(prefix)/.models
 
