diff --git a/Makefile b/Makefile
index 175ab16..c0b661b 100644
--- a/Makefile
+++ b/Makefile
@@ -92,6 +92,17 @@ $(BUILD_DIR): $(prefix)
 	mkdir -p $@
 
 
+ifdef CONFIG_DARWIN
+$(TARGET_LOADABLE): sqlite-vec.c sqlite-vec.h $(prefix)
+	$(CC) \
+		-fPIC -shared \
+		-Wall -Wextra \
+		-Ivendor/ \
+		-O3 \
+		$(CFLAGS) \
+		-install_name "@rpath/vec0.dylib" \
+		$< -o $@
+else
 $(TARGET_LOADABLE): sqlite-vec.c sqlite-vec.h $(prefix)
 	$(CC) \
 		-fPIC -shared \
@@ -100,6 +111,7 @@ $(TARGET_LOADABLE): sqlite-vec.c sqlite-vec.h $(prefix)
 		-O3 \
 		$(CFLAGS) \
 		$< -o $@
+endif
 
 $(TARGET_STATIC): sqlite-vec.c sqlite-vec.h $(prefix) $(OBJS_DIR)
 	$(CC) -Ivendor/ $(CFLAGS) -DSQLITE_CORE -DSQLITE_VEC_STATIC \
@@ -149,14 +161,8 @@ $(TARGET_CLI): sqlite-vec.h $(LIBS_DIR)/sqlite-vec.a $(LIBS_DIR)/shell.a $(LIBS_
 	examples/sqlite3-cli/core_init.c $(LIBS_DIR)/shell.a $(LIBS_DIR)/sqlite3.a $(LIBS_DIR)/sqlite-vec.a -o $@
 
 
-sqlite-vec.h: sqlite-vec.h.tmpl VERSION
-	VERSION=$(shell cat VERSION) \
-	DATE=$(shell date -r VERSION +'%FT%TZ%z') \
-	SOURCE=$(shell git log -n 1 --pretty=format:%H -- VERSION) \
-	VERSION_MAJOR=$$(echo $$VERSION | cut -d. -f1) \
-	VERSION_MINOR=$$(echo $$VERSION | cut -d. -f2) \
-	VERSION_PATCH=$$(echo $$VERSION | cut -d. -f3 | cut -d- -f1) \
-	envsubst < $< > $@
+sqlite-vec.h: sqlite-vec.h.tmpl
+	cp $< $@
 
 clean:
 	rm -rf dist
diff --git a/sqlite-vec.h.tmpl b/sqlite-vec.h.tmpl
index f49f62f..699a2f7 100644
--- a/sqlite-vec.h.tmpl
+++ b/sqlite-vec.h.tmpl
@@ -23,9 +23,9 @@
 #define SQLITE_VEC_SOURCE "${SOURCE}"
 
 
-#define SQLITE_VEC_VERSION_MAJOR ${VERSION_MAJOR}
-#define SQLITE_VEC_VERSION_MINOR ${VERSION_MINOR}
-#define SQLITE_VEC_VERSION_PATCH ${VERSION_PATCH}
+#define SQLITE_VEC_VERSION_MAJOR 0
+#define SQLITE_VEC_VERSION_MINOR 0
+#define SQLITE_VEC_VERSION_PATCH 0
 
 #ifdef __cplusplus
 extern "C" {
